File: spring-mvc-demo/src/main/java/com/btrie/faraday/config/SecurityConfig.java
package com.btrie.faraday.config;

import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.core.userdetails.User;


@EnableWebSecurity
public class SecurityConfig extends WebSecurityConfigurerAdapter {
	@Override
	protected void configure(HttpSecurity http) throws Exception {
		System.out.println("config http");
		http.csrf().disable();
		http.authorizeRequests()
			.antMatchers("/", "/public/**").permitAll()
			.antMatchers("/admin").hasRole("ADMIN")
			.anyRequest().authenticated()
			.and()
			.formLogin().loginPage("/login").permitAll()
			.and()
			.logout().logoutUrl("/logout").permitAll();
	}

	@Override
	protected void configure(AuthenticationManagerBuilder auth) throws Exception {
		System.out.println("config auth");
		auth.inMemoryAuthentication()
			.withUser(User.withDefaultPasswordEncoder().username("user").password("asdf").roles("USER"))
			.withUser(User.withDefaultPasswordEncoder().username("admin").password("asdf").roles("ADMIN"));
	}
}

File: spring-demo/server-admin/src/main/kotlin/newbros/config/SecurityConfig.kt
package newbros.config

import newbros.account.AccountService
import org.springframework.data.jpa.repository.config.EnableJpaAuditing
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.core.AuthenticationException
import org.springframework.security.crypto.password.NoOpPasswordEncoder
import javax.inject.Inject
import javax.servlet.http.HttpServletResponse
import javax.servlet.ServletException
import java.io.IOException
import org.springframework.security.web.AuthenticationEntryPoint
import javax.servlet.http.HttpServletRequest


@EnableWebSecurity
@EnableJpaAuditing
class SecurityConfig: WebSecurityConfigurerAdapter() {

	@Inject
	lateinit var accountService: AccountService

	override fun configure(auth: AuthenticationManagerBuilder) {
		auth.userDetailsService(accountService)
			// FIXME
			.passwordEncoder(NoOpPasswordEncoder.getInstance())
	}

	override fun configure(http: HttpSecurity) {
		http.csrf().disable()
			.authorizeRequests()
			.anyRequest().permitAll()
//			.anyRequest().authenticated()
			.and()
				.httpBasic()
				// suppress browser authentication dialog
				.authenticationEntryPoint(Http401AuthenticationEntryPoint("None"))
			.and()
			.logout().logoutUrl("/api/logout")
	}
}


class Http401AuthenticationEntryPoint(private val headerValue: String) : AuthenticationEntryPoint {

	@Throws(IOException::class, ServletException::class)
	override fun commence(request: HttpServletRequest, response: HttpServletResponse, authException: AuthenticationException) {
		response.setHeader("WWW-Authenticate", this.headerValue)
		response.sendError(HttpServletResponse.SC_UNAUTHORIZED, authException.message)
	}
}
